### PR TITLE
Load testing Journal

### PR DESCRIPTION
**Pull request by giorgiosironi**
Single process, which can be executed in multiple copies.

Chooses between three kind of loads to create:
- loading a static page
- going through a listing's item and pages
- searching for a random string

When a page is visited, all resources like scripts, CSS and images are accessed (through HEAD requests at the moment).

To be expanded with other pages or workflows we are interested in for stress testing.

### CODE REVIEW

File: spectrum/checks.py [hunk 1]
```diff
@@ -810,6 +813,32 @@ def _build_url(path, host):
...
+    "/about/peer-review",
+    "/alerts",
+    "/annual-reports",
+    "/archive/2016",
+    "/community",
+    "/contact",
+    "/for-the-press",
+    "/resources",
+    "/terms",
+    #/who-we-work-with
```
Comment (thewilkybarkid) [inline spectrum/checks.py]: Any reason not to include this?

File: spectrum/checks.py [hunk 2]
```diff
@@ -810,6 +813,32 @@ def _build_url(path, host):
...
+JOURNAL_GENERIC_PATHS = [
+    "/about",
+    "/about/early-career",
+    "/about/innovation",
+    "/about/openness",
+    "/about/peer-review",
+    "/alerts",
+    "/annual-reports",
+    "/archive/2016",
+    "/community",
```
Comment (thewilkybarkid) [inline spectrum/checks.py]: Is a listing page.

File: spectrum/checks.py [hunk 3]
```diff
@@ -810,6 +813,32 @@ def _build_url(path, host):
...
     host=SETTINGS['journal_cdn_host']
 )
+JOURNAL_GENERIC_PATHS = [
+    "/about",
+    "/about/early-career",
+    "/about/innovation",
+    "/about/openness",
+    "/about/peer-review",
+    "/alerts",
+    "/annual-reports",
```
Comment (thewilkybarkid) [inline spectrum/checks.py]: Is a listing page.

File: spectrum/checks.py [hunk 4]
```diff
@@ -810,6 +813,32 @@ def _build_url(path, host):
...
+    "/archive/2016",
```
Comment (thewilkybarkid) [inline spectrum/checks.py]: Is a listing of months. The page for a month is then a listing.

File: spectrum/checks.py [hunk 5]
```diff
@@ -810,6 +813,32 @@ def _build_url(path, host):
...
+    "/terms",
+    #/who-we-work-with
+]
+JOURNAL_LISTING_PATHS = [
+    '/articles/correction',
+    '/collections',
+    '/inside-elife',
+    '/labs',
+    '/podcast',
+    '/subjects',
```
Comment (thewilkybarkid) [inline spectrum/checks.py]: Each subject's page is then a listing.

File: spectrum/checks.py (hunk 1 above)
Comment (giorgiosironi) [inline spectrum/checks.py]: probably wasn't ready at the time, I'll add it

**Comment by giorgiosironi** [comment]
Ok, only error is the kitchen sink video not being streamed due to the caching error, will wait to merge (does not add load testing to builds but the infrastructure to access journal in end2end tests will do more checks now, like loading videos and similar download links with HEAD requests).

**Comment by giorgiosironi** [comment]
One process is pretty slow due to all the requests it has to perform:
```
[2017-04-20 16:26:15,219][INFO][spectrum.load][] New run
[2017-04-20 16:26:15,219][INFO][spectrum.load][] Selecting action JournalListing(/articles/correction, pages queue length 1, items queue length 0)
[2017-04-20 16:26:15,219][INFO][spectrum.load][] Loading listing page /articles/correction
[2017-04-20 16:26:23,942][INFO][spectrum.load][] New run
[2017-04-20 16:26:23,943][INFO][spectrum.load][] Selecting action JournalPage(/about)
[2017-04-20 16:26:23,943][INFO][spectrum.load][] Loading fixed page /about
[2017-04-20 16:26:35,793][INFO][spectrum.load][] New run
[2017-04-20 16:26:35,794][INFO][spectrum.load][] Selecting action JournalSearch(length=3)
[2017-04-20 16:26:35,794][INFO][spectrum.load][] Searching for yxr
[2017-04-20 16:26:47,397][INFO][spectrum.load][] New run
[2017-04-20 16:26:47,397][INFO][spectrum.load][] Selecting action JournalPage(/about/early-career)
[2017-04-20 16:26:47,397][INFO][spectrum.load][] Loading fixed page /about/early-career
[2017-04-20 16:27:01,568][INFO][spectrum.load][] New run
[2017-04-20 16:27:01,568][INFO][spectrum.load][] Selecting action JournalHomepage(links queue length 0)
[2017-04-20 16:27:01,568][INFO][spectrum.load][] Loading homepage /
[2017-04-20 16:27:21,318][INFO][spectrum.load][] New run
[2017-04-20 16:27:21,318][INFO][spectrum.load][] Selecting action JournalListing(/collections, pages queue length 1, items queue length 0)
[2017-04-20 16:27:21,318][INFO][spectrum.load][] Loading listing page /collections
[2017-04-20 16:27:44,191][INFO][spectrum.load][] New run
[2017-04-20 16:27:44,191][INFO][spectrum.load][] Selecting action JournalListing(/collections, pages queue length 1, items queue length 6)
[2017-04-20 16:27:44,192][INFO][spectrum.load][] Loading listing item /collections/8de90445/the-natural-history-of-model-organisms
[2017-04-20 16:27:58,036][INFO][spectrum.load][] New run
[2017-04-20 16:27:58,036][INFO][spectrum.load][] Selecting action JournalSearch(length=3)
[2017-04-20 16:27:58,036][INFO][spectrum.load][] Searching for wzu
[2017-04-20 16:28:09,687][INFO][spectrum.load][] New run
[2017-04-20 16:28:09,687][INFO][spectrum.load][] Selecting action JournalListing(/annual-reports, pages queue length 1, items queue length 0)
[2017-04-20 16:28:09,687][INFO][spectrum.load][] Loading listing page /annual-reports
[2017-04-20 16:28:22,385][WARNING][spectrum.load][] Need to refill queue with []
[2017-04-20 16:28:22,385][INFO][spectrum.load][] New run
[2017-04-20 16:28:22,385][INFO][spectrum.load][] Selecting action JournalSearch(length=3)
[2017-04-20 16:28:22,385][INFO][spectrum.load][] Searching for rzd
[2017-04-20 16:28:33,694][INFO][spectrum.load][] New run
[2017-04-20 16:28:33,694][INFO][spectrum.load][] Selecting action JournalPage(/about/openness)
[2017-04-20 16:28:33,694][INFO][spectrum.load][] Loading fixed page /about/openness
[2017-04-20 16:28:48,450][INFO][spectrum.load][] New run
[2017-04-20 16:28:48,450][INFO][spectrum.load][] Selecting action JournalListingOfListing(/subjects, listings queue length 0)
[2017-04-20 16:28:48,451][INFO][spectrum.load][] Loading listing of listing /subjects
[2017-04-20 16:29:05,265][INFO][spectrum.load][] New run
[2017-04-20 16:29:05,265][INFO][spectrum.load][] Selecting action JournalSearch(length=3)
[2017-04-20 16:29:05,266][INFO][spectrum.load][] Searching for xas
[2017-04-20 16:29:16,896][INFO][spectrum.load][] New run
[2017-04-20 16:29:16,896][INFO][spectrum.load][] Selecting action JournalListing(/collections, pages queue length 1, items queue length 5)
[2017-04-20 16:29:16,896][INFO][spectrum.load][] Loading listing item /collections/8e3cd4ac/early-career-researchers-a-collection-of-articles
[2017-04-20 16:29:31,775][INFO][spectrum.load][] New run
[2017-04-20 16:29:31,775][INFO][spectrum.load][] Selecting action JournalListingOfListing(/archive/2016, listings queue length 0)
[2017-04-20 16:29:31,775][INFO][spectrum.load][] Loading listing of listing /archive/2016
[2017-04-20 16:29:50,164][INFO][spectrum.load][] New run
[2017-04-20 16:29:50,164][INFO][spectrum.load][] Selecting action JournalListing(/collections, pages queue length 1, items queue length 4)
[2017-04-20 16:29:50,164][INFO][spectrum.load][] Loading listing item /collections/2d2e35c4/tropical-disease-a-selection-of-papers
[2017-04-20 16:30:09,704][INFO][spectrum.load][] New run
[2017-04-20 16:30:09,704][INFO][spectrum.load][] Selecting action JournalPage(/for-the-press)
[2017-04-20 16:30:09,704][INFO][spectrum.load][] Loading fixed page /for-the-press
[2017-04-20 16:30:23,413][INFO][spectrum.load][] New run
[2017-04-20 16:30:23,414][INFO][spectrum.load][] Selecting action JournalListing(/labs, pages queue length 1, items queue length 0)
[2017-04-20 16:30:23,414][INFO][spectrum.load][] Loading listing page /labs
[2017-04-20 16:30:42,776][WARNING][spectrum.load][] Need to refill queue with []
[2017-04-20 16:30:42,776][INFO][spectrum.load][] New run
[2017-04-20 16:30:42,776][INFO][spectrum.load][] Selecting action JournalPage(/for-the-press)
[2017-04-20 16:30:42,776][INFO][spectrum.load][] Loading fixed page /for-the-press
[2017-04-20 16:30:54,424][INFO][spectrum.load][] New run
[2017-04-20 16:30:54,424][INFO][spectrum.load][] Selecting action JournalSearch(length=3)
[2017-04-20 16:30:54,424][INFO][spectrum.load][] Searching for qsd
[2017-04-20 16:31:06,393][INFO][spectrum.load][] New run
[2017-04-20 16:31:06,393][INFO][spectrum.load][] Selecting action JournalHomepage(links queue length 16)
[2017-04-20 16:31:06,393][INFO][spectrum.load][] Loading link /articles/24869
[2017-04-20 16:32:42,674][INFO][spectrum.load][] New run
[2017-04-20 16:32:42,674][INFO][spectrum.load][] Selecting action JournalListing(/podcast, pages queue length 1, items queue length 0)
[2017-04-20 16:32:42,675][INFO][spectrum.load][] Loading listing page /podcast
[2017-04-20 16:33:06,504][INFO][spectrum.load][] New run
[2017-04-20 16:33:06,504][INFO][spectrum.load][] Selecting action JournalSearch(length=3)
[2017-04-20 16:33:06,504][INFO][spectrum.load][] Searching for kaq
[2017-04-20 16:33:18,357][INFO][spectrum.load][] New run
[2017-04-20 16:33:18,357][INFO][spectrum.load][] Selecting action JournalSearch(length=3)
[2017-04-20 16:33:18,357][INFO][spectrum.load][] Searching for jhm
[2017-04-20 16:33:30,697][INFO][spectrum.load][] New run
[2017-04-20 16:33:30,697][INFO][spectrum.load][] Selecting action JournalHomepage(links queue length 15)
[2017-04-20 16:33:30,697][INFO][spectrum.load][] Loading link /articles/24869
[2017-04-20 16:35:00,590][INFO][spectrum.load][] New run
[2017-04-20 16:35:00,590][INFO][spectrum.load][] Selecting action JournalListing(/annual-reports, pages queue length 0, items queue length 4)
[2017-04-20 16:35:00,591][INFO][spectrum.load][] Loading listing item https://2015.elifesciences.org
[2017-04-20 16:35:05,807][WARNING][spectrum.load][] Need to refill queue with []
[2017-04-20 16:35:05,807][INFO][spectrum.load][] New run
[2017-04-20 16:35:05,807][INFO][spectrum.load][] Selecting action JournalPage(/contact)
[2017-04-20 16:35:05,808][INFO][spectrum.load][] Loading fixed page /contact
[2017-04-20 16:35:19,152][INFO][spectrum.load][] New run
[2017-04-20 16:35:19,152][INFO][spectrum.load][] Selecting action JournalListing(/podcast, pages queue length 1, items queue length 8)
[2017-04-20 16:35:19,152][INFO][spectrum.load][] Loading listing item /podcast/episode37
[2017-04-20 16:35:32,757][INFO][spectrum.load][] New run
[2017-04-20 16:35:32,757][INFO][spectrum.load][] Selecting action JournalListingOfListing(/archive/2016, listings queue length 12)
[2017-04-20 16:35:32,757][INFO][spectrum.load][] Loading listing page /archive/2016/january
[2017-04-20 16:35:53,416][WARNING][spectrum.load][] Need to refill queue with []
[2017-04-20 16:35:53,416][INFO][spectrum.load][] New run
[2017-04-20 16:35:53,416][INFO][spectrum.load][] Selecting action JournalHomepage(links queue length 14)
[2017-04-20 16:35:53,416][INFO][spectrum.load][] Loading link /articles/18834
[2017-04-20 16:36:38,928][INFO][spectrum.load][] New run
[2017-04-20 16:36:38,928][INFO][spectrum.load][] Selecting action JournalPage(/about/peer-review)
[2017-04-20 16:36:38,929][INFO][spectrum.load][] Loading fixed page /about/peer-review
[2017-04-20 16:36:50,443][INFO][spectrum.load][] New run
[2017-04-20 16:36:50,444][INFO][spectrum.load][] Selecting action JournalSearch(length=3)
[2017-04-20 16:36:50,444][INFO][spectrum.load][] Loading search result /articles/12260
[2017-04-20 16:37:10,048][INFO][spectrum.load][] New run
[2017-04-20 16:37:10,049][INFO][spectrum.load][] Selecting action JournalListing(/annual-reports, pages queue length 0, items queue length 3)
[2017-04-20 16:37:10,049][INFO][spectrum.load][] Loading listing item https://2014.elifesciences.org
[2017-04-20 16:37:10,324][WARNING][spectrum.load][] Need to refill queue with []
[2017-04-20 16:37:10,324][INFO][spectrum.load][] New run
[2017-04-20 16:37:10,324][INFO][spectrum.load][] Selecting action JournalListing(/inside-elife, pages queue length 1, items queue length 0)
[2017-04-20 16:37:10,325][INFO][spectrum.load][] Loading listing page /inside-elife
[2017-04-20 16:37:22,543][INFO][spectrum.load][] New run
[2017-04-20 16:37:22,543][INFO][spectrum.load][] Selecting action JournalSearch(length=3)
[2017-04-20 16:37:22,543][INFO][spectrum.load][] Loading search result /articles/00712
[2017-04-20 16:53:44,687][INFO][spectrum.load][] New run
[2017-04-20 16:53:44,687][INFO][spectrum.load][] Selecting action JournalSearch(length=3)
[2017-04-20 16:53:44,687][INFO][spectrum.load][] Searching for axf
[2017-04-20 16:53:57,849][INFO][spectrum.load][] New run
[2017-04-20 16:53:57,849][INFO][spectrum.load][] Selecting action JournalListingOfListing(/archive/2016, listings queue length 0)
[2017-04-20 16:53:57,850][INFO][spectrum.load][] Loading listing of listing /archive/2016
[2017-04-20 16:54:09,575][INFO][spectrum.load][] New run
[2017-04-20 16:54:09,575][INFO][spectrum.load][] Selecting action JournalSearch(length=3)
[2017-04-20 16:54:09,575][INFO][spectrum.load][] Searching for rll
[2017-04-20 16:54:22,528][INFO][spectrum.load][] New run
[2017-04-20 16:54:22,528][INFO][spectrum.load][] Selecting action JournalListing(/inside-elife, pages queue length 1, items queue length 0)
[2017-04-20 16:54:22,528][INFO][spectrum.load][] Loading listing page /inside-elife
[2017-04-20 16:54:35,513][INFO][spectrum.load][] New run
[2017-04-20 16:54:35,513][INFO][spectrum.load][] Selecting action JournalPage(/who-we-work-with)
[2017-04-20 16:54:35,513][INFO][spectrum.load][] Loading fixed page /who-we-work-with
[2017-04-20 16:55:50,936][INFO][spectrum.load][] New run
[2017-04-20 16:55:50,936][INFO][spectrum.load][] Selecting action JournalSearch(length=3)
[2017-04-20 16:55:50,937][INFO][spectrum.load][] Loading search result /articles/18372
```
This however is from my machine, will improve when running inside us-east-1. I'll take a look to see if it's possible to perform parallel requests too.

**Pull request merged**